src/main/java/com/algorand/algosdk/transaction/AtomicTransactionComposer.java

```diff
@@ -0,0 +1,439 @@
+                        new byte[]{},
+                        null,
+                        null
+                ));
+                continue;
+            }
+            Response<PendingTransactionResponse> resp =
+                    client.PendingTransactionInformation(this.transactionList.get(i).txn.txID()).execute();
+            if (!resp.isSuccessful()) {
+                retList.add(new ReturnValue(
+                        null,
+                        null,
+                        null,
+                        resp.message()
+                ));
+                continue;
+            }
+
+            PendingTransactionResponse respBody = resp.body();
+            List<byte[]> logs = respBody.logs;
+            byte[] retLine = null;
+            for (int logIndex = logs.size() - 1; logIndex >= 0; logIndex--) {
+                byte[] line = logs.get(logIndex);
+                if (!checkLogRet(line))
+                    continue;
+                retLine = line;
+                break;
+            }
+            if (retLine == null) {
+                retList.add(new ReturnValue(
+                        null,
+                        null,
+                        null,
+                        "expected to find logged return value, got none"
+                ));
+                continue;
+            }
+
+            byte[] abiEncoded = Arrays.copyOfRange(retLine, 4, retLine.length);
+            Object decoded = null;
+            String parseError = null;
+            try {
+                ABIType ABIType = this.methodMap.get(i).returns.parsedType;
+                decoded = ABIType.decode(abiEncoded);
+            } catch (Exception e) {
+                parseError = e.getMessage();
+            }
+            retList.add(new ReturnValue(
+                    this.transactionList.get(i).txn.txID(),
+                    abiEncoded,
+                    decoded,
+                    parseError
+            ));
+        }
+
+        return new ExecuteResult(txInfo.confirmedRound, this.getTxIDs(), retList);
+    }
+
+    private static boolean checkLogRet(byte[] logLine) {
+        if (logLine.length < 4)
+            return false;
+        for (int i = 0; i < 4; i++) {
+            if (logLine[i] != ABI_RET_HASH[i])
+                return false;
+        }
+        return true;
+    }
+
+    public static class ExecuteResult {
+        public Long confirmedRound;
+        public List<String> txIDs;
+        public List<ReturnValue> methodResults;
+
+        public ExecuteResult(Long confirmedRound, List<String> txIDs, List<ReturnValue> methodResults) {
+            this.confirmedRound = confirmedRound;
+            this.txIDs = txIDs;
+            this.methodResults = methodResults;
+        }
+    }
+
+    public static class ReturnValue {
+        public String txID;
+        public byte[] rawValue;
+        public Object value;
+        public String parseError;
```

Comment:
Can you make this an `Exception` type again? That would be more useful, since it will contain the stacktrace of the error too